bolt-sidecar/src/api/commitments/firewall_recv.rs

```diff
@@ -0,0 +1,622 @@
+    /// NOTE: Is there a better way to avoid this monster type?
+    /// SAFETY: the `poll` implementation of this struct promptly handles these responses and
+    /// ensures this vector doesn't grow indefinitely.
+    pending_commitment_responses: FuturesUnordered<Pin<Box<PendingCommitmentResult>>>,
+    /// The collection of outgoing messages to be sent to the connected websocket server.
+    outgoing_messages: VecDeque<Message>,
+}
+
+impl CommitmentRequestProcessor {
+    pub fn new(
+        url: Url,
+        tx: mpsc::Sender<CommitmentEvent>,
+        write: SplitSink<WebSocketStream<MaybeTlsStream<TcpStream>>, Message>,
+        read: SplitStream<WebSocketStream<MaybeTlsStream<TcpStream>>>,
+        ping_rx: broadcast::Receiver<()>,
+        shutdown_rx: broadcast::Receiver<()>,
+    ) -> Self {
+        Self {
+            url,
+            api_events_tx: tx,
+            write_sink: write,
+            read_stream: read,
+            ping_rx,
+            shutdown_rx,
+            pending_commitment_responses: FuturesUnordered::new(),
+            outgoing_messages: VecDeque::new(),
+        }
+    }
+}
+
+impl Future for CommitmentRequestProcessor {
+    // The output of this future is a boolean indicating whether reconnection is needed.
+    type Output = bool;
+
+    fn poll(
+        self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Self::Output> {
+        let this = self.get_mut();
+        let rpc_url = this.url.clone();
+
+        loop {
+            let mut progress = false;
+
+            // 1. Handle incoming websocket messages from the read stream.
+            while let Poll::Ready(maybe_message) = this.read_stream.poll_next_unpin(cx) {
+                progress = true;
+
+                let Some(res_message) = maybe_message else {
+                    warn!("websocket connection with {} closed by server", rpc_url);
+                    return Poll::Ready(true);
+                };
+
+                match res_message {
+                    Ok(Message::Text(text)) => {
+                        trace!(?rpc_url, text, "received text message from websocket connection");
+                        // Create the channel to send and receive the commitment response
+                        let (tx, rx) = oneshot::channel();
+
+                        // TODO: parse the text into a commitment request
+                        let request = CommitmentRequest::Inclusion(InclusionRequest::default());
+                        let event = CommitmentEvent { request, response: tx };
+
+                        if let Err(err) = this.api_events_tx.try_send(event) {
+                            error!(?err, "failed to forward commitment event to channel");
+                        }
+
+                        // add the pending response to this buffer for later processing
+                        this.pending_commitment_responses.push(rx.boxed());
```

Comment:
You can actually just push this `rx` as is without boxing it, and replace the type above with:

```rs
pending_commitment_responses:
        FuturesUnordered<oneshot::Receiver<Result<SignedCommitment, CommitmentError>>>,
```

This will also avoid a heap allocation

Reply:
Great catch, thanks!